File: packages/typescript/service/tests/redaction.test.ts
import { describe, it, expect } from "vitest";
import { redactFailure } from "../src/redaction.js";

describe("redactFailure", () => {
  const SAMPLE_FAILURE = {
    type: "scope_insufficient",
    detail:
      "Token scope ['read'] does not include required scope 'admin:write' for capability 'dangerous.action'",
    retry: true,
    resolution: {
      action: "request_scope",
      requires: "admin:write",
      grantable_by: "org-admin@example.com",
      estimated_availability: "PT1H",
    },
  };

  it("full returns unchanged", () => {
    const result = redactFailure(SAMPLE_FAILURE, "full");
    expect(result).toEqual(SAMPLE_FAILURE);
  });

  it("reduced strips grantable_by", () => {
    const result = redactFailure(SAMPLE_FAILURE, "reduced");
    expect(result.type).toBe("scope_insufficient");
    expect(result.retry).toBe(true);
    const res = result.resolution as Record<string, unknown>;
    expect(res.grantable_by).toBeNull();
    expect(res.action).toBe("request_scope");
    expect(res.requires).toBe("admin:write");
    expect(res.estimated_availability).toBe("PT1H");
  });

  it("reduced truncates long detail", () => {
    const longDetail = "x".repeat(300);
    const failure = { ...SAMPLE_FAILURE, detail: longDetail };
    const result = redactFailure(failure, "reduced");
    expect((result.detail as string).length).toBeLessThanOrEqual(200);
  });

  it("reduced preserves short detail", () => {
    const result = redactFailure(SAMPLE_FAILURE, "reduced");
    expect(result.detail).toBe(SAMPLE_FAILURE.detail);
  });

  it("redacted uses generic detail", () => {
    const result = redactFailure(SAMPLE_FAILURE, "redacted");
    expect(result.type).toBe("scope_insufficient");
    expect(result.detail).not.toBe(SAMPLE_FAILURE.detail);
    expect(result.detail as string).not.toContain("admin:write");
  });

  it("redacted strips resolution fields", () => {
    const result = redactFailure(SAMPLE_FAILURE, "redacted");
    const res = result.resolution as Record<string, unknown>;
    expect(res.requires).toBeNull();
    expect(res.grantable_by).toBeNull();
    expect(res.estimated_availability).toBeNull();
    expect(res.action).toBe("request_scope");
  });

  it("redacted preserves retry", () => {
    const result = redactFailure(SAMPLE_FAILURE, "redacted");
    expect(result.retry).toBe(true);
  });

  it("type is never redacted", () => {
    for (const level of ["full", "reduced", "redacted"]) {
      const result = redactFailure(SAMPLE_FAILURE, level);
      expect(result.type).toBe("scope_insufficient");
    }
  });

  it("policy treated as redacted", () => {
    const result = redactFailure(SAMPLE_FAILURE, "policy");
    const redacted = redactFailure(SAMPLE_FAILURE, "redacted");
    expect(result).toEqual(redacted);
  });

  it("failure without resolution", () => {
    const failure = { type: "internal_error", detail: "Something broke", retry: false };
    const result = redactFailure(failure, "redacted");
    expect(result.type).toBe("internal_error");
    expect(result.detail).toBe("Internal error");
    expect(result.retry).toBe(false);
    expect(result.resolution).toBeUndefined();
  });
});

File: packages/go/service/redaction_test.go
package service

import "testing"

func TestRedactFailure_Full(t *testing.T) {
	failure := map[string]any{
		"type":   "scope_insufficient",
		"detail": "Need travel.book scope",
		"retry":  false,
		"resolution": map[string]any{
			"action":       "request_scope",
			"requires":     []string{"travel.book"},
			"grantable_by": "human:samir@example.com",
		},
	}
	result := RedactFailure(failure, "full")
	if result["detail"] != "Need travel.book scope" {
		t.Error("full mode should preserve detail")
	}
	res := result["resolution"].(map[string]any)
	if res["grantable_by"] != "human:samir@example.com" {
		t.Error("full mode should preserve grantable_by")
	}
}

func TestRedactFailure_Reduced(t *testing.T) {
	failure := map[string]any{
		"type":   "scope_insufficient",
		"detail": "Need travel.book scope",
		"retry":  false,
		"resolution": map[string]any{
			"action":       "request_scope",
			"requires":     []string{"travel.book"},
			"grantable_by": "human:samir@example.com",
		},
	}
	result := RedactFailure(failure, "reduced")
	if result["detail"] != "Need travel.book scope" {
		t.Error("reduced should preserve detail (under 200 chars)")
	}
	res := result["resolution"].(map[string]any)
	if res["grantable_by"] != nil {
		t.Error("reduced should null grantable_by")
	}
	if res["action"] != "request_scope" {
		t.Error("reduced should preserve action")
	}
}

func TestRedactFailure_Redacted(t *testing.T) {
	failure := map[string]any{
		"type":   "scope_insufficient",
		"detail": "Need travel.book scope with specific user context",
		"retry":  false,
		"resolution": map[string]any{
			"action":       "request_scope",
			"requires":     []string{"travel.book"},
			"grantable_by": "human:samir@example.com",
		},
	}
	result := RedactFailure(failure, "redacted")
	if result["detail"] != "Insufficient scope for this capability" {
		t.Errorf("redacted detail = %q, want generic message", result["detail"])
	}
	if result["type"] != "scope_insufficient" {
		t.Error("type should never be redacted")
	}
	if result["retry"] != false {
		t.Error("retry should never be redacted")
	}
	res := result["resolution"].(map[string]any)
	if res["action"] != "request_scope" {
		t.Error("action should never be redacted")
	}
	if res["requires"] != nil {
		t.Error("redacted should null requires")
	}
	if res["grantable_by"] != nil {
		t.Error("redacted should null grantable_by")
	}
}

func TestRedactFailure_ReducedTruncatesLongDetail(t *testing.T) {
	longDetail := ""
	for i := 0; i < 300; i++ {
		longDetail += "x"
	}
	failure := map[string]any{
		"type":   "internal_error",
		"detail": longDetail,
		"retry":  false,
	}
	result := RedactFailure(failure, "reduced")
	detail := result["detail"].(string)
	if len(detail) > 200 {
		t.Errorf("reduced should truncate detail to 200 chars, got %d", len(detail))
	}
}

func TestRedactFailure_TypeAndRetryNeverRedacted(t *testing.T) {
	failure := map[string]any{
		"type":   "token_expired",
		"detail": "Token XYZ expired at ...",
		"retry":  true,
	}
	for _, level := range []string{"full", "reduced", "redacted"} {
		result := RedactFailure(failure, level)
		if result["type"] != "token_expired" {
			t.Errorf("level=%q: type was redacted", level)
		}
		if result["retry"] != true {
			t.Errorf("level=%q: retry was redacted", level)
		}
	}
}
